File: aoc15/day_12/day_12.ml
let input = Yojson.Basic.from_file "input" in

let is_red (_, v) =
  let open Yojson.Basic.Util in
  try to_string v = "red" with Type_error _ -> false
in

let rec sum nored json =
  match json with
  | `Int i -> i
  | `List l -> List.fold_left ( + ) 0 (List.map (sum nored) l)
  | `Assoc a ->
      if nored && List.exists is_red a then 0
      else
        List.fold_left ( + ) 0
          (List.map (sum nored) (List.map (fun (_, v) -> v) a))
  | _ -> 0
in

let first = sum false input in
let second = sum true input in
Printf.printf "First: %d\n" first;
Printf.printf "Second: %d\n" second

File: aoc15/go/day_12/day_12.go
package main

import (
	"aoc/collections"
	"encoding/json"
	"fmt"
	"os"
)

func sum(obj interface{}, nored bool) (res float64) {
	switch val := obj.(type) {
	case float64:
		return val
	case []interface{}:
		for _, e := range val {
			res += sum(e, nored)
		}
	case map[string]interface{}:
		if nored && collections.MapContains(val, "red") {
			return 0
		}
		for _, v := range val {
			res += sum(v, nored)
		}
	default:
	}
	return
}

func main() {
	input, _ := os.ReadFile("input")
	var doc []interface{}

	json.Unmarshal(input, &doc)

	fmt.Println("First:", int(sum(doc, false)))
	fmt.Println("Second:", int(sum(doc, true)))
}
